lib/shared/bucketing-assembly-script/assembly/types/dvcEvent.ts

```diff
@@ -0,0 +1,213 @@
+import { JSON } from 'assemblyscript-json/assembly'
+import {
+    getDateFromJSONOptional,
+    getF64FromJSONOptional,
+    getJSONObjFromJSONOptional,
+    getStringFromJSON,
+    getStringFromJSONOptional,
+    getStringMapFromJSONObj,
+    jsonArrFromValueArray,
+    jsonObjFromMap
+} from '../helpers/jsonHelpers'
+import { DVCPopulatedUser } from './dvcUser'
+import uuid from 'as-uuid/assembly'
+
+interface DVCUserInterface {
+    type: string
+    date: Date | null
+    target: string | null
+    value: f64
+    metaData: JSON.Obj | null
+}
+
+export const EventTypes = new Set<string>()
+EventTypes.add('variableEvaluated')
+EventTypes.add('aggVariableEvaluated')
+EventTypes.add('variableDefaulted')
+EventTypes.add('aggVariableEvaluated')
+
+export class DVCEvent extends JSON.Value implements DVCUserInterface {
+    constructor(
+        /**
+         * type of the event
+         */
+        public type: string,
+
+        /**
+         * target / subject of event. Contextual to event type
+         */
+        public target: string | null,
+
+        /**
+         * date event occurred according to client stored as time since epoch
+         */
+        public date: Date | null,
+
+        /**
+         * value for numerical events. Contextual to event type
+         */
+        public value: f64,
+
+        /**
+         * extra metadata for event. Contextual to event type
+         */
+        public metaData: JSON.Obj | null
+    ) {
+        super()
+    }
+
+    static fromJSONString(eventStr: string): DVCEvent {
+        const eventJSON = JSON.parse(eventStr)
+        if (!eventJSON.isObj) throw new Error('DVCEvent eventStr not a JSON Object')
+        const event = eventJSON as JSON.Obj
+
+        return new DVCEvent(
+            getStringFromJSON(event, 'type'),
+            getStringFromJSONOptional(event, 'target'),
+            getDateFromJSONOptional(event, 'date'),
+            getF64FromJSONOptional(event, 'value', NaN),
+            getJSONObjFromJSONOptional(event, 'metaData')
+        )
+    }
+
+    stringify(): string {
+        const json = new JSON.Obj()
+        json.set('type', this.type)
+        const date = this.date // nullable check only works like this
+        if (date) json.set('date', date.toISOString())
+        if (this.target) json.set('target', this.target)
+        if (!isNaN(this.value)) json.set('value', this.value)
+        if (this.metaData) json.set('metaData', this.metaData)
+        return json.stringify()
+    }
+}
+
+export class DVCRequestEvent extends JSON.Value {
```

Comment:
perhaps document in a comment what this class is for and how its different from DVCEvent